packages/liveblocks-yjs/src/index.ts

```diff
@@ -0,0 +1,253 @@
+  ) {
+    this.doc = doc;
+    this.room = room;
+
+    // if we have a connectionId already during construction, use that
+    const connectionId = this.room.getSelf()?.connectionId;
+    if (connectionId) {
+      this.doc.clientID = connectionId;
+    }
+    this.awareness = new Awareness(this.doc, this.room);
+    this.doc.on("update", this.updateHandler);
+
+    this.unsubscribers.push(
+      this.room.events.connection.subscribe((e) => {
+        if (e === "open") {
+          this.syncDoc();
+        }
+      })
+    );
+
+    this.unsubscribers.push(
+      this.room.events.docUpdated.subscribe((update) => {
+        Y.applyUpdate(this.doc, Base64.toUint8Array(update), "backend");
+      })
+    );
+
+    if (config?.httpEndpoint) {
+      this.httpEndpoint = config.httpEndpoint + "?room=" + this.room.id;
+
+      this.unsubscribers.push(
+        this.room.events.customEvent.subscribe(({ event }) => {
+          if ((event as RoomEvent)?.type === "REFRESH") {
+            void this.resyncHttp();
+          }
+        })
+      );
+
+      void this.resyncHttp();
+    }
+    this.syncDoc();
+  }
+
+  private syncDoc = () => {
+    /**
+     * If the connection changes, set the new id, this is used by awareness.
+     * yjs' only requirement for clientID is that it's truly unique and a number.
+     * Liveblock's connectionID satisfies those constraints
+     *  */
+    this.doc.clientID = this.room.getSelf()?.connectionId || this.doc.clientID;
+    this.awareness.clientID = this.doc.clientID; // tell our awareness provider the new ID
+
+    // The state vector is sent to the server so it knows what to send back
+    // if you don't send it, it returns everything
+    const encodedVector = Base64.fromUint8Array(Y.encodeStateVector(this.doc));
+    this.room.getYDoc(encodedVector);
+  };
+
+  private updateHandler = async (update: Uint8Array, origin: string) => {
+    if (origin !== "backend") {
+      const encodedUpdate = Base64.fromUint8Array(update);
+      this.room.updateYDoc(encodedUpdate);
+      if (this.httpEndpoint) {
+        await fetch(this.httpEndpoint, {
+          method: "POST",
+          body: encodedUpdate,
+        });
+      }
+    }
+  };
+
+  private async resyncHttp() {
+    if (!this.httpEndpoint) {
+      return;
+    }
+    const response = await fetch(
+      `${this.httpEndpoint}${
+        this.lastUpdateDate !== null
+          ? `&after=${this.lastUpdateDate.toISOString()}`
+          : ""
+      }`
+    );
+    const { updates, lastUpdate } = (await response.json()) as RefreshResponse;
+
+    if (updates.length === 0) {
+      return;
+    }
+
+    this.lastUpdateDate = new Date(lastUpdate);
+
+    const update = Y.mergeUpdates(updates.map(Base64.toUint8Array));
+    Y.applyUpdate(this.doc, update, "backend");
+  }
+
+  destroy(): void {
+    this.doc.off("update", this.updateHandler);
+    this.unsubscribers.forEach((unsub) => unsub());
+    this.awareness.destroy();
+  }
```

Comment:
> You can see in my tiptap example I'm always cleaning it up in the useEffect return.

Oh, I missed that cleanup code (didn't pay enough attention to the examples on my previous review round). Perhaps a silly question, but can you destroy a doc before you destroy the provider that relies on that doc? Should those destroy calls happen in reverse order?

I wonder if we could somehow structure the API to be a bit more like how we automatically enter/leave the room on mount/unmount. Perhaps it's not possible with this API, because the Doc and Provider are created by the user directly, and therefore cleanup must also be managed by the user directly 🤔 

Maybe not a problem if people are used to do that working with Yjs already indeed.

> _…but a lot of times people don't notice they have multiple Ydocs floating around in memory being sync'd at once because they actually work pretty well_

Ah yes, I can see that 😂

Reply:
I think if we implemented an editor on top of this, like a LiveblocksLexical or something, then yes we can handle it, otherwise I think this falls on the user.